lib/rouge/lexers/bbcbasic.rb

```diff
@@ -0,0 +1,115 @@
+# -*- coding: utf-8 -*- #
+# frozen_string_literal: true
+
+module Rouge
+  module Lexers
+    class BBCBASIC < RegexLexer
+      title "BBCBASIC"
+      desc "BBC BASIC syntax"
+      tag 'bbcbasic'
+      filenames '*,fd1'
+
+      def self.punctuation
+        @punctuation ||= %w(
+          [:,;'~] SPC TAB
+        )
+      end
+
+      def self.function
+        @function ||= %w(
+          ABS ACS ADVAL ASC ASN ATN BEATS BEAT BGET# CHR\$ COS COUNT DEG EOF#
+          ERL ERR ERROR EVAL EXP EXT# GET GET\$ GET\$# HIMEM INKEY INKEY\$
+          INSTR INT LEFT\$ LEN LN LOG LOMEM MID\$ OPENIN OPENOUT OPENUP PAGE
+          POS PTR# RAD REPORT\$ RIGHT\$ RND SGN SIN SQR STR\$ STRING\$ SUM
+          SUMLEN TAN TEMPO TIME TIME\$ TOP USR VAL VPOS
+        )
+      end
+
+      def self.control
+        @control ||= %w(
+          CASE CHAIN ELSE ENDCASE ENDIF ENDPROC ENDWHILE END FN FOR GOSUB GOTO
+          IF INSTALL LIBRARY NEXT OF OTHERWISE OVERLAY PROC REPEAT RETURN STEP
+          STOP THEN TO UNTIL WHEN WHILE
+        )
+      end
+
+      def self.statement
+        @statement ||= %w(
+          BEATS BPUT# CALL CLEAR CLG CLOSE# CLS COLOR COLOUR DATA DIM ENVELOPE
+          GCOL LET MODE OFF ON ORIGIN OSCI PLOT PRINT PRINT# QUIT READ REPORT
+          SOUND STEREO SWAP SYS TINT VDU VOICE VOICES WAIT WIDTH
+        )
+      end
+
+      def self.operator
+        @operator ||= %w(
+          << <= <> < >= >>> >> > [-!\$()*+\/=?^|] AND DIV EOR MOD NOT OR
+        )
+      end
+
+      def self.constant
+        @constant ||= %w(
+          FALSE TRUE
+        )
+      end
+
+      state :root do
+        rule %r/[ \n]+/, Text
+        rule %r/[\[]/, Keyword, :assembly1
+        rule %r/\*.*/, Generic::Prompt # CLI command
+        rule %r/REM *>.*/, Comment::Special
+        rule %r/REM.*/, Comment
+        rule %r/#{BBCBASIC.punctuation.join('|')}/, Punctuation
+        rule %r/#{BBCBASIC.function.join('|')}/, Name::Builtin # function or pseudo-variable
```

Comment:
Should you put a `\b` at the end here to avoid inadvertent matches with names that begin with a matching word?

Reply:
Although maybe this would cause problems with function names that end in a `#`.

Reply:
Sorry, I'm jumping around a bit as I review this.

Given how much interpolation you want to do, I think it might improve performance to match a generic pattern and pass it a block inside of which you do different things depending on whether the captured text matches an element of one of the arrays.

You can see an example of this being done in the JavaScript lexer:

https://github.com/rouge-ruby/rouge/blob/7cdb1f9da96660afb144568c8b9029e474797ec9/lib/rouge/lexers/javascript.rb#L176-L192

Reply:
`\b` isn't that useful for this language because, for historical reasons, back-tick is valid in variable and user-defined function names, and as you note, `$` and `#` appear in a number of built-in keywords too.

As it happpens, the BASIC interpreter will lex strings that begin with keywords in the same way that this regex code does, e.g. `PRINTCOSY` becomes `PRINT` `COS` `Y`, or `LETS=1` becomes `LET` `S` `=` `1`.